README
This is the README for the Rasberry Pi Ham Hadio Repeater Controller
project.  The Goal of this project is to develop an inexpensive
linking repeater controller witha cost of aproximatly $100/port vs
current commerical solutions which are closer to $160/port.

This is an extensable linkable controller with both local and remote
linking.

For information on the mmra network see mmra.org
For information on this project contact jml@jmlzone.com.

For information on setting up this project on a Pi3 please see the Quickstart.txt

Most of the code will be in Python, however some of it is in C.

Eventually there will be good documeation and a bug/enhancement
tracking WIKI or some such thing.

Current state: The test repeater is stable on the air but many many
things are still to be done.

----------------------------------------------------------------------
Mailing lists for the project
  Please join one or more of the groups below.
MMRA-PiPtr-Hardware+subscribe@groups.io
MMRA-PiPtr-Software+subscribe@groups.io
MMRA-PiPtr-Documentation+subscribe@groups.io

----------------------------------------------------------------------

As you know the MMRA maintains a network of 22 repeaters.  A repeater
is basically a receiver, a transmitter and a controller
in-between. While the network generally works, there are some upgrades
we are continually working on.  One of these efforts being led by me,
James N1DDK, is developing a new repeater controller to enhance the
remote controllability, monitoring and adjusting of the
repeaters. Most of the repeater controller in use today are based on
designs over 20 years old. There have been many advances in technology
in the past 20 years. This project is bringing a more modern, more
scalable and more software based and controllable approach based on
mostly standard hardware and using popular modern methods.

 There are some guiding principles to this project:
1) No Jumpers, switches or mechanical pots that can be set incorectly at a site needing a return visit.
2) Open source -- its part of the Ham culture to experiment and share. This will also ensure longevity of the project beyond the initial developers.
3) Better ease of use than similar commercial products.
4) Scaleability and cost per port more effective than the commercial products.
5) Mostly software for simple functions like CTCSS encode / decode, DTMF encode / decode, timers, morse etc.
6) Well documented

Goals
 * Ease of use for users and control operators.
 * Web back end to view status and control linking.
 * Ability to sell hardware as a fundraiser for MMRA.

To this end I have been working mostly solo with some input from K1IW and W1BRI (and some non members) and I have a working prototype on the air on 224.080 as part of a test platform. The next goals will be to make some boards and have further testing and development.

Here is a list of some of the help that is desired:
 * Programming help in C and Python (possibly java/java script for some of the web interface)
 * Schematics design/review for the custom audio board.
 * Documentation help -- there is some basic documentation, but it would be good to have much more on both hardware, software and user guides.
 * Mechanical/sheet metal design for an enclosure, etc
 * Building help, both mechanical and building the initial boards for testing.
 * Testing help using the controller on a bench and trying things before we put them on the air.
 * On the air testing.

Want to learn more?
 There is a git-hub repository for the software, documentation and issue tracking:
https://github.com/jmlzone/PiPtr
There may be a talk about the project at Boxborough!

Even if you don't know anything about software or repeater controllers you can still help! You don't need to know anything to be a documentation volunteer!

Ready to help?
Contact James, n1ddk@mmra.org.



notes/history.txt
    1  uptime
    2  ls /dev/usb
    3  ls /dev/usb/hiddev0 
    4  sudo reboot
    5  echo $AUDIODEV
    6  ls /dev/dsp
    7  aplay -1
    8  aplay -l
    9  ls /dev
   10  man aplay
   11  sudo apt-get update
   12  sudo apt-get upgrade
   13  sudo raspi-config
   14  exit
   15  which git
   16  pwd
   17  cd python
   18  ls
   19  git clone git://github.com/swharden/Python-GUI-examples
   20  ls
   21  cd Python-GUI-examples/
   22  ls
   23  cd 2016-07-37_qt_audio_monitor/
   24  ls
   25  python go.py 
   26  sudo apt-get install PyQt4
   27  hg clone https://www.riverbankcomputing.com/hg/sip
   28  which hg
   29  pip3 install SIP
   30  wget https://sourceforge.net/projects/pyqt/files/sip/sip-4.18.1/sip-4.18.1.tar.gz
   31  ls
   32  tar tfz sip-4.18.1.tar.gz 
   33  tar xfz sip-4.18.1.tar.gz
   34  cd sip-4.18.1/
   35  ls
   36  more README 
   37  python build.py prepare
   38  ls
   39  more README 
   40  ls
   41  more configure.py 
   42  python configure.py -help
   43  sudo python configure.py
   44  ls -ltr
   45  sudo make
   46  ls
   47  more Makefile 
   48  make 
   49  pwd
   50  cd ..
   51  wget http://sourceforge.net/projects/pyqt/files/PyQt4/PyQt-4.11.4/PyQt-x11-gpl-4.11.4.tar.gz
   52  ls
   53  ls PyQt-x11-gpl-4.11.4.tar.gz 
   54  tar tfz PyQt-x11-gpl-4.11.4.tar.gz 
   55  ls
   56  cd ../
   57  ls
   58  cd 2016-07-37_qt_audio_monitor/
   59  ls
   60  more ui_main.py 
   61  ls
   62  more SWHear.py 
   63  sudo apt-get install python-dev
   64  sudo apt-get install -y python-dev --fix-missing
   65  sudo apt-get update
   66  sudo apt-get install -y python-dev --fix-missing
   67  pwd
   68  ls
   69  cd sip-4.18.1/
   70  ls
   71  make
   72  make clean
   73  make
   74  sudo make installl
   75  sudo make install
   76  ls
   77  cd ../
   78  ls
   79  tar xfz PyQt-x11-gpl-4.11.4.tar.gz 
   80  cd PyQt-x11-gpl-4.11.4
   81  ls
   82  more README 
   83  python configure-ng.py
   84  sudo apt-get install qt
   85  sudo apt-get install qt4-dev-tools
   86  python configure-ng.py
   87  history
   88  ls
   89  more README 
   90  make 
   91  sudo make install
   92  ls
   93  cd ../
   94  ls
   95  python go.py 
   96  history
   97  python go.py 
   98  git clone https://github.com/pyqtgraph/pyqtgraph
   99  ls
  100  cd pyqtgraph/
  101  ls
  102  python setup.py
  103  python setup.py -help
  104  python setup.py --help
  105  python setup.py build
  106  sudo python setup.py install
  107  cd ../
  108  ls
  109  python go.py
  110  hi
  111  history
  112  python go.py
  113  git clone http://people.csail.mit.edu/hubert/git/pyaudio.git
  114  cd pyaudio/
  115  ls
  116  sudo apt-get install python-pyaudio
  117  python go.py
  118  pwd
  119  cd ../
  120  python go.py
  121  history
  122  sudo apt-get install python-matlibplot
  123  sudo apt-get install matlibplot
  124  sudo apt-get install matplotlib
  125  sudo apt-get install python-matplotlib
  126  python go.py
  127  sudo apt-get install python-scipy
  128  python go.py
  129  xterm &
  130  emacs 
  131  cd c
  132  ls
  133  cd multimon/
  134  ls
  135  ./bin-armv7l/multimon 
  136  ./bin-armv7l/multimon scope
  137  ls
  138  more unixinput.c 
  139  ls
  140  more demod_afsk12.c 
  141  ls
  142  more demod_dtmf.c 
  143  ls
  144  more mkcostab.c 
  145  ls
  146  ls bin-armv7l/mkcostab
  147  bin-armv7l/mkcostab
  148  ls -ltr
  149  more costab.c
  150  more costabf.c 
  151  more costabi.c 
  152  ls
  153  emacs Makefile
  154  bg
  155  ls
  156  more gen.c 
  157  ls
  158  more xdisplay.c 
  159  ls
  160  more multimon.h 
  161  cd python
  162  ls
  163  cd Python-GUI-examples/
  164  ls
  165  cd 2016-07-37_qt_audio_monitor/
  166  puthon go.py 
  167  python go.py 
  168  pwd
  169  cd python
  170  ls
  171  pwd
  172  ls
  173  ls ../
  174  mv ../gistfile1.py .
  175  more .
  176  more gistfile1.py 
  177  which python
  178  python gistfile1.py 
  179  python my_goertzel.py 
  180  python
  181  cd pythin
  182  cd python
  183  ls
  184  mroe my_goertzel.py 
  185  more my_goertzel.py 
  186  python my_goertzel.py
  187  ls ~/.ssh
  188  mkdir ~/.ssh
  189  ifconfig | more
  190  ls -l ~/.ssh
  191  python my_goertzel.py
  192  python setup.py build_ext -- inplace
  193  python setup.py build_ext --inplace
  194  pip instal Cython
  195  sudo pip install Cython
  196  python setup.py build_ext --inplace
  197  ls -ltr
  198  python my_goertzel.py
  199  cd ../
  200  mkdir c
  201  cd d
  202  cd c
  203  ls
  204  cc -o alsa_listen alsa_listen.c
  205  ls /usr/lib/alsa
  206  ls /usr/lib
  207  sudo apt-get install libasound2-dev
  208  sudo apt-get -y install libasound2-dev
  209  cc -o alsa_listen alsa_listen.c
  210  cc -o alsa_listen alsa_listen.c -lasound
  211  ls
  212  ./alsa_listen 
  213  which xgps
  214  sudo apt-get install gpsd gpsd-clients python-gps
  215  sudo lsusb
  216  ls /dev/tty*
  217  sudo gpsd /dev/ttyUSB0 -F /var/run/gpsd.sock
  218  cgps -s
  219  xgps
  220  sudo lsusb
  221  cgps
  222  sudo reboot
  223  sudo lsusb
  224  ls /dev/tty*
  225  man gpsd
  226  gpsd /dev/ttyUSB0
  227  cgps
  228  mau gpsd
  229  man gpsd
  230  gpsd /dev/ttyUSB0 -D2
  231  ps -auxww | grep gps
  232  kill 1046
  233  sudo kill 1046
  234  ps -auxww | grep gps
  235  man gpsd
  236  gpsd -D 2 -N /dev/ttyUSB0 
  237  minicom /dev/ttyUSB0 
  238  sudo apt-get install -Y minicom
  239  sudo apt-get install -y minicom
  240  minicom
  241  sudo lsusb
  242  ls /dev/usb
  243  ls /dev/usb/hiddev0 
  244  ls /dev/tty*
  245  gpsd -D 2 -N /dev/ttyUSB0 
  246  sudu shutdown -h now
  247  sudo shutdown -h now
  248  minicom /dev/ttyUSB0 
  249  sudo minicom /dev/ttyUSB0 
  250  ls -l /dev/ttyUSB0 
  251  sudo chmod a+rw /dev/ttyUSB0
  252  ls -l /dev/ttyUSB0 
  253  echo hello >> /dev/ttyUSB0
  254  man stty
  255  more .asoundrc
  256  more /usr/share/alsa/alsa.conf
  257  more /etc/asound.conf
  258  aplay /usr/share/sounds/GNUstep/Ping.wav 

c/mout.c
#include "mout.h"
#include <stdio.h>
#include <stdlib.h>
#include <alsa/asoundlib.h>

const unsigned char mlookup[] = {115  /*,--..--*/ ,1   /*-*/ ,
  106 /*..-.-.-*/ ,41  /*/-..-.*/ ,63  /*0-----*/ ,62  /*1.----*/ ,
  60  /*2..---*/ ,56  /*3...--*/ ,48  /*4....-*/ ,32  /*5.....*/ ,
  33  /*6-....*/ ,35  /*7--...*/ ,39  /*8---..*/ ,47  /*9----.*/ ,
  1   /*:*/ ,1   /*;*/ ,1   /*<*/ ,49  /*=-...-*/ ,1   /*>*/ ,
  76  /*?..--..*/ ,1   /*@*/ ,6   /*A.-*/ ,17  /*B-...*/ ,21  /*C-.-.*/ ,
  9   /*D-..*/ ,2   /*E.*/ ,20  /*F..-.*/ ,11  /*G--.*/ ,16  /*H....*/ ,
  4   /*I..*/ ,30  /*J.---*/ ,13  /*K-.-*/ ,18  /*L.-..*/ ,7   /*M--*/ ,
  5   /*N-.*/ ,15  /*O---*/ ,22  /*P.--.*/ ,27  /*Q--.-*/ ,10  /*R.-.*/ ,
  8   /*S...*/ ,3   /*T-*/ ,12  /*U..-*/ ,24  /*V...-*/ ,14  /*W.--*/ ,
  25  /*X-..-*/ ,29  /*Y-.--*/ ,19  /*Z--..*/};

int elesamples;
#define DIT elesamples
#define DAH DIT*3
#define WSPACE DIT*6
#define BUFLEN 128
#define CHAN 1
int ch_idx;
int ph;
int phinc;
int samples;
int ampl;
short buf[BUFLEN * CHAN];
snd_pcm_t *playback_handle;
int buf_idx;
int err;

/*----------------------------------------------------------------------
  mout Morse Output 
----------------------------------------------------------------------*/
void mout(char * msg){
  int i = 0;
  int cindex;
  int mchar;
  while (msg[i] != 0) {
    cindex = msg[i] & 0x7f; //remove any parity
    if(cindex <= 'z'){
      if(cindex >= 'a') cindex = cindex-0x20; // make upper case
      cindex = cindex - ','; // comma is zero based in the table
      if(cindex <0) { // a space
	gen_space(WSPACE);
      } else { // send the charater from the lookup table
	mchar = mlookup[cindex];
	while(mchar !=1) {
	  //ON
	  if(mchar & 0x01) {// test lsb
	    gen_tone(DAH);
	  } else {
	    gen_tone(DIT);
	  }
	  //OFF
	  gen_space(DIT);
	  mchar = mchar >>1; // shift down the bits
	}
      }
    }
    gen_space(DAH);
    i++; // move to next char
  }
}
void write_buf(void){
    if ((err = snd_pcm_writei (playback_handle, buf, BUFLEN)) != BUFLEN) {
      fprintf (stderr, "write to audio interface failed (%s)\n",
	       snd_strerror (err));
      exit(EXIT_FAILURE);
    }
}
void send_sample(int sample) {
  int i;
  for (i=0; i<CHAN; i++){
    buf[buf_idx++] = sample;
    if( buf_idx == BUFLEN) {
      write_buf();
      buf_idx=0;
    }
  }
}
void gen_tone(int samples){
  int i;
  for (i=0; i<samples; i++){
    send_sample((ampl * COS(ph)) >> 15);
    ph += phinc;
  }
}

void gen_space(int samples){
  int i;
  for (i=0; i<samples; i++){
    send_sample(0);
  }
}

int main (int argc, char *argv[]) {
  int wpm;
  int toneFreq;
  char *msg;
  if(argc != 6) {
    printf("must use exactly 6 arguments %d given\n", argc);
    printf("%s         device        wpm freq amplitude message (in quotes)\n",argv[0]);
    printf("%s         default       22  660  10000     \"test \"\n",argv[0]);
    return EXIT_FAILURE;
  }

  wpm = atoi(argv[2]);
  toneFreq = atoi(argv[3]);
  ampl = atoi(argv[4]);
  msg = argv[5];
  unsigned int sampleRate = 22050;
  float elelen = 92.31 * 13.0 / (float) wpm;
  elesamples = (int) ((float) sampleRate * elelen /1000.0);
  int err;
  snd_pcm_hw_params_t *hw_params;

  buf_idx=0; 
  ph = 0;
  phinc = (float)0x10000 * toneFreq / sampleRate;
  //  printf("trying to open %s\n", argv[1]);
  if ((err = snd_pcm_open (&playback_handle, argv[1], SND_PCM_STREAM_PLAYBACK, 0)) < 0) {
    fprintf (stderr, "cannot open audio device %s (%s)\n", 
	     argv[1],
	     snd_strerror (err));
    return EXIT_FAILURE;
  }

  if ((err = snd_pcm_hw_params_malloc (&hw_params)) < 0) {
    fprintf (stderr, "cannot allocate hardware parameter structure (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
				 
  if ((err = snd_pcm_hw_params_any (playback_handle, hw_params)) < 0) {
    fprintf (stderr, "cannot initialize hardware parameter structure (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
	
  if ((err = snd_pcm_hw_params_set_access (playback_handle, hw_params, SND_PCM_ACCESS_RW_INTERLEAVED)) < 0) {
    fprintf (stderr, "cannot set access type (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }

  if ((err = snd_pcm_hw_params_set_format (playback_handle, hw_params, SND_PCM_FORMAT_S16_LE)) < 0) {
    fprintf (stderr, "cannot set sample format (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
	
  if ((err = snd_pcm_hw_params_set_rate_near (playback_handle, hw_params, &sampleRate, 0)) < 0) {
    fprintf (stderr, "cannot set sample rate (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
	
  if ((err = snd_pcm_hw_params_set_channels (playback_handle, hw_params, CHAN)) < 0) {
    fprintf (stderr, "cannot set channel count (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
	
  if ((err = snd_pcm_hw_params (playback_handle, hw_params)) < 0) {
    fprintf (stderr, "cannot set parameters (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
	
  snd_pcm_hw_params_free (hw_params);
	
  if ((err = snd_pcm_prepare (playback_handle)) < 0) {
    fprintf (stderr, "cannot prepare audio interface for use (%s)\n",
	     snd_strerror (err));
    return EXIT_FAILURE;
  }
  printf("SampleRate = %d, wpm = %d eleSamples = %d\n", sampleRate, wpm, elesamples);
	
  mout(msg);
  if(buf_idx != 0) {
    gen_space(BUFLEN-buf_idx);
  }
  snd_pcm_drain (playback_handle);
  snd_pcm_close (playback_handle);
  return EXIT_SUCCESS;
  
}
